arangod/Aql/OptimizerRules.cpp

```diff
@@ -4057,33 +4155,62 @@ class RestrictToSingleShardChecker final : public WalkerWorker<ExecutionNode> {
       case EN::TRAVERSAL:
       case EN::SHORTEST_PATH: {
         _stop = true;
-        return true; // abort enumerating, we are done already!
+        return true;  // abort enumerating, we are done already!
+      }
+
+      case EN::FILTER: {
+        auto collectionVariables =
+            ExecutionNode::castTo<FilterNode const*>(en)->collectionVariables();
+        for (auto pair : collectionVariables) {
+          aql::Collection const* collection = pair.second;
+          aql::Variable const* collectionVariable = pair.first;
+          std::string shardId =
+              getSingleShardId(_plan, en, collection, collectionVariable);
+          if (shardId.empty() && _shardsUsed[collection].empty()) {
+            _shardsUsed[collection].emplace("all");
+          } else if (!shardId.empty()) {
+            if (1 == _shardsUsed[collection].size() &&
```

Comment:
I am not sure if this is safe like this.
What happens if we use the same collection twice in a query, e.g.
```
FOR doc1 IN collection
  FOR doc2 IN collection
    FILTER doc2.shardKey == @whatever
...
```
If this restricts the query to a single shard, it would be wrong. The reason is that the inner loop can be restricted to one shard, but the outer loop (for the same collection) must not. Is this somehow handled by the state machine?